optuna/integration/dask.py

```diff
@@ -0,0 +1,733 @@
+import asyncio
+import datetime
+import inspect
+from typing import Any
+from typing import Callable
+from typing import Container
+from typing import Dict
+from typing import Generator
+from typing import Iterable
+from typing import List
+from typing import Optional
+from typing import Sequence
+from typing import Tuple
+from typing import Union
+import uuid
+
+import optuna
+from optuna._experimental import experimental_class
+from optuna._imports import try_import
+from optuna.distributions import BaseDistribution
+from optuna.distributions import distribution_to_json
+from optuna.distributions import json_to_distribution
+from optuna.storages import BaseStorage
+from optuna.study import StudyDirection
+from optuna.study._frozen import FrozenStudy
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+with try_import() as _imports:
+    import distributed
+    from distributed.utils import thread_state
+    from distributed.worker import get_client
+
+# Serialization utilities
+
+
+def _serialize_datetime(dt: datetime.datetime) -> dict:
+    return {"__datetime__": True, "as_str": dt.strftime("%Y%m%dT%H:%M:%S.%f")}
+
+
+def _deserialize_datetime(data: dict) -> datetime.datetime:
+    return datetime.datetime.strptime(data["as_str"], "%Y%m%dT%H:%M:%S.%f")
+
+
+def _serialize_trialstate(state: TrialState) -> str:
+    return state.name
+
+
+def _deserialize_trialstate(state: str) -> TrialState:
+    return getattr(TrialState, state)
+
+
+def _serialize_frozentrial(trial: FrozenTrial) -> dict:
+    data = trial.__dict__.copy()
+    data["state"] = _serialize_trialstate(data["state"])
+    attrs = [a for a in data.keys() if a.startswith("_")]
+    for attr in attrs:
+        data[attr[1:]] = data.pop(attr)
+    data["distributions"] = {k: distribution_to_json(v) for k, v in data["distributions"].items()}
+    if data["datetime_start"] is not None:
+        data["datetime_start"] = _serialize_datetime(data["datetime_start"])
+    if data["datetime_complete"] is not None:
+        data["datetime_complete"] = _serialize_datetime(data["datetime_complete"])
+    data["value"] = None
+    return data
+
+
+def _deserialize_frozentrial(data: dict) -> FrozenTrial:
+    data["state"] = _deserialize_trialstate(data["state"])
```

Comment:
Can we just use `TrialState` like the following instead of `_deserialize_trialstate`?

```suggestion
    data["state"] = TrialState(data["state"])
```

Reply:
It looks that we can also remove `_deserialize_studydirection`.

Reply:
Sure, pushed a commit to do so